include/common/core/allocator.hpp

```diff
@@ -0,0 +1,84 @@
+#include <algorithm>
+#include <array>
+#include <variant>
+
+#include "common/core/logging.h"
+
+/*
+ * A generic pool allocator class that can be used to statically allocate memory
+ * for typically dynamic objects (such as maps).
+ *
+ * Example Use-Case:
+ *
+ *
+ */
+
+template <typename Element>
+struct AllocatorElement {
+    // The address of element member must be the same as the address of an
+    // AllocatorElement.
+    // Thus, element must be first member of this struct. And there can be no
+    // member functions and no private variables.
+    // See https://en.cppreference.com/w/cpp/types/is_standard_layout
+    Element element{};
+    bool used = false;
+};
+
+// should have type of object and max number of.
+// should have backing be an array and/or pointer to array. Do not use malloc!
+// only support allocating memory for one type of object at a time.
+
+// Use concept to check that individual size
+template <typename Element, std::size_t MaxElements>
+class PoolAllocator {
+  public:
+    using value_type = Element;
+    using size_type = std::size_t;
+    using difference_type = ptrdiff_t;
+    using pointer = Element*;
+    using const_pointer = const Element*;
+    using reference = Element&;
+    using const_reference = const Element&;
+
+    PoolAllocator() {}
+    auto operator=(PoolAllocator&) -> PoolAllocator& = delete;
+    auto operator=(PoolAllocator&&) -> PoolAllocator&& = delete;
+    PoolAllocator(const PoolAllocator&) = default;
+    PoolAllocator(PoolAllocator&&) = delete;
+    ~PoolAllocator() = default;
+
+    void deallocate(pointer loc, size_t) {
+        // Find index into backing array using the fact that loc is the first
+        // item in the AllocatorElement struct and so has the same address.
+        auto index = (reinterpret_cast<AllocatorElement<Element>*>(loc) -
```

Comment:
You also need to handle unaligned deletes here with something like
```
auto wrapped = reinterpret_cast<AllocatorElement<Element>*>(loc);
if (wrapped < backing.begin()) return;
if ((wrapped - backing.begin()) % sizeof(*wrapped) != 0) return;
```
this shouldn't ever happen, but if something called `deallocate` with a wrongly-aligned address, then you wouldn't necessarily set an arbitrary byte, but you'd set the previous element which might or might not be right.